src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala

```diff
@@ -222,6 +218,28 @@ class WorkspaceService(protected val argUserToken: WithAccessToken, val rawlsDAO
 
   }
 
+  def unPublishSuccessMessage(workspaceNamespace: String, workspaceName: String): String = s" The workspace $workspaceNamespace:$workspaceName has been un-published."
+
+  def deleteWorkspace(ns: String, name: String): Future[PerRequestMessage] = {
+    rawlsDAO.getWorkspace(ns, name) flatMap { ws =>
+      val wsPublished: Boolean = isPublished(ws)
+      if (wsPublished) {
+        setWorkspacePublishedStatus(ws.workspace, publishArg = false, rawlsDAO, ontologyDAO, searchDAO) flatMap { updatedWs =>
+          rawlsDAO.deleteWorkspace(ns, name) map { wsResponse =>
+            RequestComplete(wsResponse.copy(message = Some(wsResponse.message.getOrElse("") + unPublishSuccessMessage(ns, name))))
+          }
+        } recover {
+          case e: FireCloudExceptionWithErrorReport => RequestComplete(e.errorReport.statusCode.getOrElse(InternalServerError), ErrorReport(message = s"You cannot delete this workspace: ${e.errorReport.message}"))
+          case e: Throwable => RequestComplete(InternalServerError, ErrorReport(message = s"You cannot delete this workspace: ${e.getMessage}"))
+        }
+      } else {
+        rawlsDAO.deleteWorkspace(ws.workspace.namespace, ws.workspace.name) map { response =>
+          RequestComplete(response)
+        }
+      }
+    }
+  }
```

Comment:
The other way I've seen this kind of logic flow handled is:
```
      val unpublishFuture = if (isPublished(ws))
        setWorkspacePublishedStatus(ws.workspace, publishArg = false, rawlsDAO, ontologyDAO, searchDAO)
      else
        Future.successful(ws)

      unpublishFuture flatMap { _ =>
        rawlsDAO.deleteWorkspace(ns, name) map { wsResponse =>
          RequestComplete(wsResponse.copy(message = Some(wsResponse.message.getOrElse("") + unPublishSuccessMessage(ns, name))))
        } recover {
          case e: FireCloudExceptionWithErrorReport => RequestComplete(e.errorReport.statusCode.getOrElse(InternalServerError), ErrorReport(message = s"You cannot delete this workspace: ${e.errorReport.message}"))
          case e: Throwable => RequestComplete(InternalServerError, ErrorReport(message = s"You cannot delete this workspace: ${e.getMessage}"))
        }
      }
```
I think it's slightly cleaner, but I'm not officially requesting a change for this PR; your call if you want to adopt different syntax or not.

Reply:
👍  I like that approach better, more granular and less nest-y.